include/range/v3/at.hpp

```diff
@@ -14,31 +15,83 @@
 #ifndef RANGES_V3_AT_HPP
 #define RANGES_V3_AT_HPP
 
+#include <stdexcept>
 #include <range/v3/range_fwd.hpp>
 #include <range/v3/begin_end.hpp>
 #include <range/v3/range_traits.hpp>
 #include <range/v3/range_concepts.hpp>
-#include <range/v3/utility/iterator.hpp>
+#include <range/v3/distance.hpp>
 #include <range/v3/utility/static_const.hpp>
+#include <range/v3/index.hpp>
 
 namespace ranges
 {
     inline namespace v3
     {
+        /// Checked indexed range access.
+        ///
         /// \ingroup group-core
         struct at_fn
         {
             /// \return `begin(rng)[n]`
             template<typename Rng,
-                CONCEPT_REQUIRES_(RandomAccessRange<Rng>())>
+                CONCEPT_REQUIRES_(RandomAccessRange<Rng>() && SizedRange<Rng>())>
             RANGES_CXX14_CONSTEXPR
-            auto operator()(Rng &&rng, range_difference_type_t<Rng> n) const ->
+            auto operator()(Rng &&rng, range_difference_type_t<Rng> n) const 
+                noexcept(noexcept(begin(rng)[n]) && noexcept(ranges::distance(rng)))
+              ->
                 decltype(begin(rng)[n])
             {
+                if (n >= ranges::distance(rng))
+                {
+                    throw std::out_of_range("ranges::at");
+                }
                 return begin(rng)[n];
             }
+            /// \return `begin(rng)[n]`
+            template<typename Rng,
+                CONCEPT_REQUIRES_(RandomAccessRange<Rng>() && !SizedRange<Rng>())>
+            RANGES_DEPRECATED(
+                "Checked indexed range access (ranges::at) on !SizedRanges is deprecated! "
+                "This version performs unchecked access (the range size cannot be computed in O(1) for !SizedRanges)! "
+                "Use ranges::index for unchecked access instead!")
+            RANGES_CXX14_CONSTEXPR
+            auto operator()(Rng &&rng, range_difference_type_t<Rng> n) const
```

Comment:
@ericniebler are `index` and `at` customization points? AFAIK only `begin` and `end` are, so IIUC this doesn't matter.